internal/binding/local.go

```diff
@@ -0,0 +1,226 @@
+// Copyright 2023 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package binding
+
+import (
+	"context"
+	"flag"
+	"fmt"
+	"net"
+	"net/netip"
+	"path/filepath"
+	"time"
+
+	"github.com/google/gopacket"
+	"github.com/google/uuid"
+	"github.com/openconfig/ondatra/binding"
+	"google.golang.org/grpc/credentials/local"
+
+	"github.com/openconfig/lemming"
+	"github.com/openconfig/lemming/dataplane/dplaneopts"
+	"github.com/openconfig/lemming/dataplane/forwarding/infra/fwdcontext"
+	"github.com/openconfig/lemming/dataplane/forwarding/util/queue"
+
+	opb "github.com/openconfig/ondatra/proto"
+
+	saipb "github.com/openconfig/lemming/dataplane/proto"
+	fwdpb "github.com/openconfig/lemming/proto/forwarding"
+)
+
+type LocalBind struct {
+	binding.Binding
+}
+
+// Local is a local (in-process) binding for lemming and magna.
+func Local(topoDir string) func() (binding.Binding, error) {
+	dir, _ := filepath.Abs(topoDir)
+	testbedFile := filepath.Join(dir, "testbed.pb.txt")
+
+	flag.Set("testbed", testbedFile)
+	return func() (binding.Binding, error) {
+		return &LocalBind{}, nil
+	}
+}
+
+type localLemming struct {
+	binding.AbstractDUT
+	l     *lemming.Device
+	dutID string
+}
+
+const (
+	gnmiPort      = 9339
+	gribiPort     = 9559
+	bgpPort       = 1179
```

Comment:
For this local binding, are they all running in the same network namespace? If so should this be configurable since each lemming will spin up a BGP server?

Reply:
they are running on different IPs